accord-core/src/main/java/accord/api/MessageSink.java

```diff
@@ -27,6 +28,6 @@
 public interface MessageSink
 {
     void send(Id to, Request request);
-    void send(Id to, Request request, Callback callback);
+    void send(Id to, Request request, CommandStore commandStore, Callback callback);
```

Comment:
Should `CommandStore` be added to the callback interface?

Reply:
I felt that this way was cleaner for the following reason: callbacks do not normally pick a CommandStore, so adding there would require us to plumb this through

Reply:
after running `BurnTest` for a long time I found I actually have to use `Executor` as some things are run in `CommandStore` only because I forced them to with this change... stuff like our test epoch population logic shouldn't be running in a CommandStore

> Maybe make the storage for Callback intrusive so it doesn't need to be a separate allocation?
> It would take a little jiggering for it to be a base class since it is currently an interface.

I don't follow this comment, what do you mean?

Reply:
`SafeCallback` just stores the `CommandStore` that `Callback` will need. So why not put the storage in `Callback` and make it an abstract base class. Then you save an allocation for `SafeCallback`.

Reply:
> So why not put the storage in Callback and make it an abstract base class.

This has an annoying impact to all code, as all code must now figure out what store to run in rather than the small set that need to do this today.  ATM this is inferred in the majority of cases, so only a few cases need to think about this.

Reply:
I am only suggesting putting the storage in the base class. Setting the value of the `CommandStore` would still be done the same way.

Reply:
@aweisberg and I spoke in slack, here is the jist of the convo...

We have `AccordCallback` in C* that translates C* callbacks to Accord callback
We have 3 different networking implementations in accord testing (maelstrom, MockCluster, BurnTest)

This means we have 4 different places that need to implement callbacks correctly, so need to centralize this logic to avoid copy/paste duplication; this is what `SafeCallback` does.


Now, we "could" move the `Executor` pointer to `Callback` and make this `abstract` (which @aweisberg was proposing) but this doesn't lower memory costs, and now adds mutable state into the system.

We could also have a `executor()` method, but would need to update all implementations... at the money this is figured out when you try to send a message, so this inference is centralized; I feel there is more value to centralize the inference than having each callback try

Reply:
I didn't realize that you always have to allocate a wrapper for `MessagingService' to adapt to the different callback interface. So there is nothing to be gained by moving stuff into `Callback`.